pkg/resource/firewall/hook.go

```diff
@@ -27,3 +36,251 @@ var (
+}
+
+// compareLoggingDestinationConfigs generates LogDestinationConfigs which need to be
+// added and deleted to reach desired LogDestinationConfig from latest LogDestinationConfig
+func compareLoggingDestinationConfigs(desired []*svcapitypes.LogDestinationConfig, latest []*svcapitypes.LogDestinationConfig) (add []*svcapitypes.LogDestinationConfig, delete []*svcapitypes.LogDestinationConfig) {
+	add = make([]*svcapitypes.LogDestinationConfig, 0)
+	delete = make([]*svcapitypes.LogDestinationConfig, 0)
+
+	desiredMap := makeLogDestinationConfigMap(desired)
+	latestMap := makeLogDestinationConfigMap(latest)
+
+	for key, val := range desiredMap {
+		if _, ok := latestMap[key]; !ok {
+			add = append(add, val)
+		}
+	}
+
+	for key, val := range latestMap {
+		if _, ok := desiredMap[key]; !ok {
+			delete = append(delete, val)
+		}
+	}
+
+	return add, delete
+}
+
+// syncLoggingConfiguration gets desired logging config and latest (existing)
+// logging config as input. It compares both and applies the delta to ensure
+// desired logging config is configured for the firewall.
+func (rm *resourceManager) syncLoggingConfiguration(
+	ctx context.Context,
+	desired *resource,
+	latest *resource,
+) (err error) {
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.syncLoggingConfiguration")
+	defer exit(err)
+
+	var input *svcsdk.UpdateLoggingConfigurationInput
+	var add, delete []*svcapitypes.LogDestinationConfig
+	if latest != nil && desired != nil {
+		add, delete = compareLoggingDestinationConfigs(desired.ko.Spec.LoggingConfiguration.LogDestinationConfigs, latest.ko.Spec.LoggingConfiguration.LogDestinationConfigs)
+		input = rm.newLoggingConfigurationPayload(latest)
+	} else if latest == nil {
+		add = desired.ko.Spec.LoggingConfiguration.LogDestinationConfigs
+		input = rm.newLoggingConfigurationPayload(desired)
+		input.LoggingConfiguration = &svcsdk.LoggingConfiguration{}
+	} else if desired == nil {
+		delete = latest.ko.Spec.LoggingConfiguration.LogDestinationConfigs
+		input = rm.newLoggingConfigurationPayload(latest)
+	}
+
+	// UpdateLoggingConfiguration allows only single LogDestinationConfig
+	// update at a time. So the updates (delete/add) need to be done in a loop.
+	for _, c := range delete {
+		resf0elem := &svcsdk.LogDestinationConfig{}
+
+		resf0elem.SetLogDestination(c.LogDestination)
+		resf0elem.SetLogDestinationType(*c.LogDestinationType)
+		resf0elem.SetLogType(*c.LogType)
+
+		for i, config := range input.LoggingConfiguration.LogDestinationConfigs {
+			if reflect.DeepEqual(config, resf0elem) {
+				// Swap with the last element
+				input.LoggingConfiguration.LogDestinationConfigs[i] = input.LoggingConfiguration.LogDestinationConfigs[len(input.LoggingConfiguration.LogDestinationConfigs)-1]
+				// Reduce the slice's length by 1
+				input.LoggingConfiguration.LogDestinationConfigs = input.LoggingConfiguration.LogDestinationConfigs[:len(input.LoggingConfiguration.LogDestinationConfigs)-1]
+				break
```

Comment:
Thinking out loud: can we put the API Call here and do not rely a break statement?

Reply:
done